File: algorithms/code/leetcode/lc155_min_stack/LC155MinStack.java
package leetcode.lc155_min_stack;

import java.util.PriorityQueue;
import java.util.Stack;

/**
 * https://leetcode.com/problems/min-stack/ #easy
 */
public final class LC155MinStack {
    private LC155MinStack() {
    }

    public static final class MinStack {
        private Stack<Integer> stack = new Stack<>();
        private PriorityQueue<Integer> heap = new PriorityQueue<>();

        public void push(final int x) {
            stack.push(x);
            heap.add(x);
        }

        public void pop() {
            heap.remove(stack.pop());
        }

        public int top() {
            return stack.peek();
        }

        public int getMin() {
            return heap.peek();
        }
    }
}
package leetcode.lc155_min_stack;

import org.junit.jupiter.api.Test;

import leetcode.lc155_min_stack.LC155MinStack.MinStack;

import static org.junit.jupiter.api.Assertions.assertEquals;

public final class LC155MinStackTests {
    @Test
    public void testExample() throws Exception {
        MinStack solution = new MinStack();
        solution.push(5);
        assertEquals(5, solution.getMin());
        solution.push(4);
        assertEquals(4, solution.getMin());
        solution.pop();
        assertEquals(5, solution.getMin());
        solution.push(3);
        assertEquals(3, solution.getMin());
        solution.top();
        assertEquals(3, solution.getMin());
        solution.push(2);
        assertEquals(2, solution.getMin());
        solution.push(1);
        assertEquals(1, solution.getMin());
    }
}

File: algorithms/code/leetcode/lc155_min_stack/lc155_min_stack.py
# https://leetcode.com/problems/min-stack/

import unittest


class MinStack:
    def __init__(self) -> None:
        self.stack: list[int] = []
        self.min_stack: list[int] = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self) -> None:
        if not self.stack:
            return
        if self.stack[-1] == self.min_stack[-1]:
            self.min_stack.pop()
        self.stack.pop()

    def top(self) -> int:
        if not self.stack:
            return -1
        return self.stack[-1]

    def getMin(self) -> int:
        if not self.min_stack:
            return -1
        return self.min_stack[-1]


class TestCode(unittest.TestCase):
    def test_examle(self) -> None:
        solution = MinStack()
        solution.pop()
        assert solution.top() == -1
        assert solution.getMin() == -1
        solution.push(5)
        assert solution.getMin() == 5
        solution.push(4)
        assert solution.getMin() == 4
        solution.pop()
        assert solution.getMin() == 5
        solution.push(3)
        assert solution.getMin() == 3
        solution.top()
        assert solution.getMin() == 3
        solution.push(2)
        assert solution.getMin() == 2
        solution.push(1)
        assert solution.getMin() == 1
